edi-module/component_writer.bal

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2023 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+isolated function writeComponentGroup(json componentGroup, EDISegSchema segSchema, EDIFieldSchema fieldSchema, EDIContext context) returns string|Error {
+    if !(componentGroup is map<json> || componentGroup is () || componentGroup is string) {
```

Comment:
Seems this validation logic is common for `writeComponentGroup` and `writeSubcomponentGroup` functions. So we can generalize it to method as follows.

```ballerina
isolated function checkValidity(json ediMessage, EDISegSchema segSchema, EDIComponentSchema|EDIFieldSchema schema) returns string|Error? {
    if !(ediMessage is map<json> || ediMessage is () || ediMessage is string) {
        return error Error(string `Input field is not compatible with the schema. 
            Segment schema: ${printSegMap(segSchema)}, Component field schema: ${schema.toString()}, Input ${ediMessage.toString()}`);
    }
    if ediMessage is () {
        if schema.required {
            return error Error(string `Mandatory field is not found in the input.
                Segment ${segSchema.tag}, Field: ${schema.tag}`);
        }
        return "";
    }
    if ediMessage is string {
        if ediMessage.trim() != "" {
            return error Error(string `Input field is not compatible with the schema. 
                Segment schema: ${printSegMap(segSchema)}, Composite field schema: ${schema.toString()}, Input ${ediMessage.toString()}`);
        }
        if schema.required {
            return error Error(string `Mandatory field is not found in the input.
                Segment ${segSchema.tag}, Field: ${schema.tag}`);
        }
        return "";
    }
    return ();
}
```

Then within each function we can invoke that method as follows.

```ballerina
isolated function writeComponentGroup(json componentGroup, EDISegSchema segSchema, EDIFieldSchema fieldSchema, EDIContext context) returns string|Error {
    string|Error? checkValidityResult = checkValidity(componentGroup, segSchema, fieldSchema);
    if checkValidityResult !is () {
        return checkValidityResult;
    }
    string cd = context.schema.delimiters.component;
    string[] ckeys = (<map<json>>componentGroup).keys();
    ....... 
```

This will reduce the code lines and simplifies the logic. The only catch is we will have to cast the edi message to `<map<json>` in few places when accessing (refer to last line above), as it is not type checked within this function it self. @MaryamZi is there a way to avoid that casting in this approach?

Reply:
Let's delay this until this casting problem is solved. In addition, we can give more specific error messages if keep the logic separate.